### PR TITLE
Fortran training example

### LINKED ISSUE #446: Training a model in fortran
Hello,
I was wondering if it is possible to enable the training of a model directly in fortran when imported from a torch-script.
Have you any plan in order to do so?

In particular:
- I want to use my optimizer in fortran, and in order to do so I would like to have a callback in fortran for getting the 
  parameters of the model, then change them accordingly to my optimizer.

Any idea, or also a suggestion is highly appreciated.

Thank you,
Nicola

### PR DESCRIPTION
**Pull request by joewallwork**
Closes #446.

This PR demonstrates training a model in Fortran using an optimizer. Good old SimpleNet is retrained to be able to permute an input array, shifting by one entry. This is demonstrated to work, given some fairly arbitrary training parameters. Here's a loss curve I was able to plot:

losses

To enable this, we need to be able to access the model parameters to pass to the model so some source changes were required. See https://docs.pytorch.org/cppdocs/frontend.html#end-to-end-example for an end-to-end example in C++.

## Checklist

- [x] Add README
- [x] Docs
- [x] Drop Python version as not needed
- [x] Better explain what we are trying to retrain the model to do.
- [x] Tests for the source code changes
- [x] Changelog entry.

### CODE REVIEW

File: src/ctorch.cpp [hunk 1]
```diff
@@ -792,6 +810,28 @@ bool torch_jit_module_is_training(const torch_jit_script_module_t module) {
...
+  }
+  int i = 0;
+  for (auto parameter : parameters) {
+    // Allocate a new heap-based Tensor that is a shallow copy (alias) of the
+    // model parameter.  PyTorch's copy constructor shares underlying storage,
+    // so the optimizer will update the model's actual weights through this
+    // handle.  Writing the address back into out[i] initialises the previously
+    // unset Fortran torch_tensor variable, matching the pattern used by every
+    // other tensor-creating function (torch_empty, torch_from_blob, etc.).
+    out[i] = new torch::Tensor(parameter);
```
Comment (joewallwork) [inline src/ctorch.cpp:833]: In the Fortran wrapper, we might want to check this isn't associated beforehand.

File: src/ftorch_model.f90 [hunk 2]
```diff
@@ -184,6 +184,41 @@ end function torch_jit_model_is_training_c
     is_training = torch_jit_model_is_training_c(self%p)
   end function torch_model_is_training
 
+  !> Extracts the parameters from a model
+  subroutine torch_model_get_parameters(model, output_tensors)
```
Comment (joewallwork) [inline src/ftorch_model.f90]: Could this be a function that returns `output_tensors`? It doesn't need constructing beforehand.

That way we could make it a type-bound function, too.

Comment (joewallwork) [reply src/ftorch_model.f90]: It turns out you can't do this with assumed shape.

File: src/ctorch.cpp (hunk 1 above)
Comment (joewallwork) [reply src/ctorch.cpp:833]: Done in https://github.com/Cambridge-ICCS/FTorch/pull/581/commits/d0416c861812627a8444b1e9d03c77027e3a17db.

File: examples/09_Autograd/simplenet_backward.f90
Comment (joewallwork) [inline examples/09_Autograd/simplenet_backward.f90:1]: This example needed fixing following the changes underpinning `torch_tensor_forward`. I got a bit carried away and also added code demonstrating differentiation w.r.t. weights. I'd be happy to take that out into a separate PR if preferred.

File: src/ctorch.cpp [hunk 3]
```diff
@@ -762,11 +762,32 @@ void torch_jit_module_forward(const torch_jit_script_module_t module,
...
       // Multiple output models will return a tuple => cast to tensors.
       for (int i = 0; i < nout; ++i) {
-        std::move(*out[i]) = model_out.toTuple()->elements()[i].toTensor();
+        auto tensor = model_out.toTuple()->elements()[i].toTensor();
+        // Same two strategies as the single-tensor branch above.
+        if (requires_grad) {
+          *out[i] = tensor;
+        } else {
+          std::move(*out[i]) = tensor;
+        }
```
Comment (joewallwork) [inline src/ctorch.cpp:790]: This change breaks the API in the case `requires_grad=true` because any association with Fortran data is lost. I couldn't see a way to avoid it but maybe something can be done?

**Comment by jatkinson1000** [comment]
I'm getting a unit test failure on Mac for model parameters requires grad, same as the CI:

<details>

```console
test 13
      Start 13: unittest_model_parameters

13: Test command: /Users/jwa34/FTorch_main/build_ftorch/test/unit/model/unittest_model_parameters "--verbose"
13: Working Directory: /Users/jwa34/FTorch_main/build_ftorch/test/unit/model
13: Test timeout computed to be: 10000000
13:
13:
13:  Start: <unittest_model_parameters_suite.test_parameters_values[F][F]>
13: .PASSED :: [test_parameters_values] relative tolerance =  0.1000E-04
13:    end: <unittest_model_parameters_suite.test_parameters_values[F][F]>
13:
13:
13:  Start: <unittest_model_parameters_suite.test_parameters_requires_grad[F][F]>
13: . Failure in <unittest_model_parameters_suite.test_parameters_requires_grad[F][F]>
13: F   end: <unittest_model_parameters_suite.test_parameters_requires_grad[F][F]>
13:
13:
13:  Start: <unittest_model_parameters_suite.test_parameters_requires_grad[T][T]>
13: .   end: <unittest_model_parameters_suite.test_parameters_requires_grad[T][T]>
13:
13: Time:         0.010 seconds
13:
13: Failure
13:  in:
13: unittest_model_parameters_suite.test_parameters_requires_grad[F][F]
13:   Location:
13: [unittest_model_parameters.pf:135]
13: AssertEquivalent failure:
13:     Expected: <F>
13:     Actual:   <T>
13:
13:  FAILURES!!!
13: Tests run: 3, Failures: 1, Errors: 0
13: , Disabled: 0
13: STOP *** Encountered 1 or more failures/errors during testing. ***
13/17 Test #13: unittest_model_parameters ..................***Failed  Error regular expression found in output. Regex=[Encountered 1 or more failures/errors during testing]  0.15 sec
```

</details>

File: examples/09_Autograd/simplenet_backward.f90
Comment (jatkinson1000) [reply examples/09_Autograd/simplenet_backward.f90:1]: Here is fine I think.
They're not huge and a knock-on effect.

File: examples/09_Autograd/README.md
Comment (jatkinson1000) [inline examples/09_Autograd/README.md:1]: Changes to this exercise generally look fine. Does it need any addition to the text to state that it now also outputs the gradient with respect to the weights, and what/why we expected these to be?

File: examples/11_Training/plot.py
Comment (jatkinson1000) [inline examples/11_Training/plot.py:1]: Could add a title if I'm being pedantic.

File: examples/11_Training/README.md [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+This can be represented by the weights matrix
+```
+[0, 0, 0, 0, 1]
+[1, 0, 0, 0, 0]
+[0, 1, 0, 0, 0]
+[0, 0, 1, 0, 0]
+[0, 0, 0, 1, 0]
+```
+In each iteration of the training loop, we generate a random input vector and
+apply the permutation to get the expected output.
```
Comment (jatkinson1000) [inline examples/11_Training/README.md]: Suggestion for more explicit statement:
```suggestion
manually apply the permutation to get the expected output as training data
before conducting a training step.
```

File: examples/11_Training/README.md [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+can be run with:
+```sh
+python3 plot.py
+```
+This will read the data files and produce a `losses.png` file in the current
+directory. The result should look something like
+
+![Losses](example_losses.png)
+
+The use of random numbers in the training means that the plot you generate will
```
Comment (jatkinson1000) [inline examples/11_Training/README.md:172]: I wonder if there is an interface to set the random seed for C++.
Definitely for a follow-up PR though.

File: src/ctorch.cpp (hunk 3 above)
Comment (jatkinson1000) [reply src/ctorch.cpp:790]: I think this is probably OK.
To our knowledge no-one is actively using this yet?

I'd also be happy to make the next release v2.0 as it includes a significant change to pt2ts.

File: src/ctorch.cpp [hunk 6]
```diff
@@ -762,11 +762,32 @@ void torch_jit_module_forward(const torch_jit_script_module_t module,
     auto model_out = model->forward(inputs_vec);
     if (model_out.isTensor()) {
       // Single output models will return a tensor directly.
-      std::move(*out[0]) = model_out.toTensor();
+      // Two assignment strategies are used depending on whether gradient
+      // tracking is needed:
+      //   requires_grad=true (training):
+      //     Plain lvalue assignment selects Tensor::operator=(Tensor&&) &
+      //     This is a shallow swap that leaves the grad_fn intact so that we can
+      //     differentiate with respect to weights
```
Comment (jatkinson1000) [inline src/ctorch.cpp:770]: Is it possible to inspect/use the values of the intermediate output Tensor from Fortran in this case, or do we leave it as an intermediate `torch_tensor` never associated with a Fortran array?

Comment (jatkinson1000) [reply src/ctorch.cpp:770]: Ah, the note in the Fortran source answers this 😄

File: src/ftorch_tensor.fypp [hunk 7]
```diff
@@ -1080,13 +1080,14 @@ contains
...
     end interface
 
-    if (tensor%get_rank() /= 1) then
-      write(*,*) "Error :: external gradient can only be implicitly created for scalar fields"
-      stop 1
-    end if
-
-    sizes(:) = tensor%get_shape()
-    if (sizes(1) /= 1) then
+    ! Accept a rank-0 tensor (0-dim PyTorch scalar) or a rank-1 size-1 tensor
```
Comment (jatkinson1000) [inline src/ftorch_tensor.fypp]: Took me a moment to untangle the logic here 😅
A break definitely needed!

File: test/unit/model/unittest_model_parameters.pf [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    integer(c_int64_t), parameter :: tensor_shape(ndims) = [5]
+
+    ! Load the test model from file
+    call torch_model_load(model, trim(filename), device_type, device_index, this%param%requires_grad)
+
+    ! Extract the model parameters
+    call torch_model_parameters(model, parameters_tensors)
+
+    ! Check that the parameters have the correct requires_grad property
+    ! FIXME: When requires_grad=.false. we get .true.. Is this right?
```
Comment (jatkinson1000) [inline test/unit/model/unittest_model_parameters.pf]: Just highlighting that this is still here!
I think you already know as I believe this is what's failing tests.

File: examples/11_Training/training.f90 [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+  call torch_tensor_from_array(in_tensors(1), in_data, torch_kCPU)
+  call torch_tensor_from_array(out_tensors(1), out_data, torch_kCPU)
+  call torch_tensor_from_array(target_tensors(1), target_data, torch_kCPU)
+
+  ! Load ML model with training enabled
+  call torch_model_load(model, args(1), torch_kCPU, is_training=.true.)
+
+  ! Get weights from model
+  ! NOTE: We don't construct weights_tensors because we want it to point to weights data created by
+  !       the model constructor
```
Comment (jatkinson1000) [inline examples/11_Training/training.f90]:
```suggestion
  ! NOTE: We don't construct weights_tensors from a Fortran array because we want it
  !        to point to weights data created by the model constructor
```

File: examples/11_Training/training.f90 [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+    call random_number(in_data)
+    target_data(:) = in_data([5,1,2,3,4])
+
+    ! Forward pass: run inference
+    call torch_model_forward(model, in_tensors, out_tensors, requires_grad=.true.)
+
+    ! Evaluate the loss function as computed mean square error (MSE) between target and input, then
+    ! log its value
+    ! NOTE: We reconstruct loss each iteration so torch_tensor_mean's rank/shape check
+    !       sees a fresh rank-1 size-1 tensor.  torch_tensor_mean then replaces the impl.
```
Comment (jatkinson1000) [inline examples/11_Training/training.f90]: Clarify "`impl`"?

File: examples/11_Training/training.f90
Comment (jatkinson1000) [inline examples/11_Training/training.f90:1]: This all makes sense to me.
The only comment I think I have is that it seems some things in here are for illustrative purposes in the exercise, but would not necessarily required in practice (though may be very useful to debug in practice).

Some of these are clear, but others are less so, so I wonder if it's possible to add a note in the README or top of file to be clear about this.

This might be overkill though, so happy for you to disagree! 😅

**Review by jatkinson1000 — COMMENTED** [review]
Hey @joewallwork worked through this and generally all looking good to me.

Left a couple of comments, but all pretty minor.
I'll leave approval until tests pass and you are happy with the hanging fixme.

Note that there is a [known issue](https://github.com/fortran-lang/setup-fortran/issues/215) with `setup-fortran` causing it to fail. Hopefully this will be resolved in a few days... 🤞

File: examples/09_Autograd/README.md
Comment (joewallwork) [reply examples/09_Autograd/README.md:1]: Simplified the example and added some explanation in c438a3d. Was nice to work through the maths.

File: examples/11_Training/plot.py
Comment (joewallwork) [reply examples/11_Training/plot.py:1]: Added in 386cd3099542ab65fff3ad8468b32119ca3c475f.

File: examples/11_Training/README.md (hunk 5 above)
Comment (joewallwork) [reply examples/11_Training/README.md:172]: See https://discuss.pytorch.org/t/seeding-a-torch-generator-in-c/185437/3. Someone comments there that the result isn't consistent with Python, though.

File: examples/09_Autograd/README.md
Comment (joewallwork) [reply examples/09_Autograd/README.md:1]: Ah, fixed in 5d87d1e.

File: test/unit/model/unittest_model_parameters.pf (hunk 8 above)
Comment (joewallwork) [reply test/unit/model/unittest_model_parameters.pf]: Yeah, am working on it.

File: examples/11_Training/training.f90
Comment (joewallwork) [reply examples/11_Training/training.f90:1]: Which things were you thinking of in particular?

File: src/ftorch_tensor.fypp (hunk 7 above)
Comment (joewallwork) [reply src/ftorch_tensor.fypp]: Hopefully 4a2d5db is a bit clearer?

File: examples/11_Training/training.f90 (hunk 10 above)
Comment (joewallwork) [reply examples/11_Training/training.f90]: It turns out that note was out of date so I dropped it entirely in 4b2615b. Also made some of the other comments clearer.

File: src/ctorch.cpp (hunk 3 above)
Comment (joewallwork) [reply src/ctorch.cpp:790]: Okay cool. I guess not really related to this PR but personally I don't think the pt2ts change necessarily implies API change - after all, the previous approach was just to provide some templates.

File: test/unit/model/unittest_model_parameters.pf (hunk 8 above)
Comment (joewallwork) [reply test/unit/model/unittest_model_parameters.pf]: Further investigation didn't really help. It's unclear from the C++ docs what should happen here. I removed the test and opened #585.

Comment (jatkinson1000) [reply test/unit/model/unittest_model_parameters.pf]: Huh, I guess we can leave if here for now then as the rest of the code seems to empirically work.

File: examples/11_Training/training.f90
Comment (jatkinson1000) [reply examples/11_Training/training.f90:1]: I think things like an additional inference pass to populate data for output, or constructing tensors to hold gradients for output that would not normally be used.

However, on re-reading I feel these are fairly well commented, so feel free to ignore this if you are happy.

File: CHANGELOG.md [hunk 11]
```diff
@@ -15,6 +15,9 @@ For specific details see the [FTorch online documentation](https://cambridge-icc
   [#555](https://github.com/Cambridge-ICCS/FTorch/pull/555)
 - Added a worked example for training in Fortran in
   [#581](https://github.com/Cambridge-ICCS/FTorch/pull/581)
+- Ability to build only integration or unit tests as desired using CMake options in
+  [#584](https://github.com/Cambridge-ICCS/FTorch/pull/584)
+
```
Comment (jatkinson1000) [inline CHANGELOG.md:20]: Seems odd that this appears here, perhaps as a result of merging main in?

File: examples/09_Autograd/simplenet_backward.py [hunk 12]
```diff
@@ -26,19 +26,13 @@ def forward(self, batch: torch.Tensor) -> torch.Tensor:
...
 print(f"dy/dx = {dydx}")
 assert torch.allclose(dydx, 2.0 * external_gradient)
 
-y.backward(external_gradient)
-
 weights = model._fwd_seq[0].weight
 dydw = weights.grad
 print(f"dy/d(weights):\n{dydw}")
-assert torch.allclose(dydw, torch.vstack(5 * [x]))
+assert torch.allclose(dydw, torch.ones((5, 1)) @ x)
```
Comment (jatkinson1000) [inline examples/09_Autograd/simplenet_backward.py:38]: Nice use of `@` 😉

**Review by jatkinson1000 — APPROVED** [review]
Thanks @joewallwork this all looks great to me now, thanks for responding to the comments.

Left a couple open, but feel free to resolve and merge once you are happy! 🚀

File: CHANGELOG.md (hunk 11 above)
Comment (joewallwork) [reply CHANGELOG.md:20]: Yeah I think so - in the main files changed tab it only shows the entry that I actually added for this PR.

Comment (jatkinson1000) [reply CHANGELOG.md:20]: Ah, yes, I was using "changes since last reviewed"!
